src/Control/Middleware/HTTPCacheControlMiddleware.php

```diff
@@ -0,0 +1,557 @@
+<?php
+
+namespace SilverStripe\Control\Middleware;
+
+use InvalidArgumentException;
+use SilverStripe\Control\HTTP;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\HTTPResponse_Exception;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Core\Resettable;
+
+class HTTPCacheControlMiddleware implements HTTPMiddleware, Resettable
+{
+    use Configurable;
+    use Injectable;
+
+    const STATE_ENABLED = 'enabled';
+
+    const STATE_PUBLIC = 'public';
+
+    const STATE_PRIVATE = 'private';
+
+    const STATE_DISABLED = 'disabled';
+
+    /**
+     * Generate response for the given request
+     *
+     * @todo Refactor HTTP::add_cache_headers() (e.g. etag handling) into this middleware
+     *
+     * @param HTTPRequest $request
+     * @param callable $delegate
+     * @return HTTPResponse
+     */
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        try {
+            $response = $delegate($request);
+        } catch (HTTPResponse_Exception $ex) {
+            $response = $ex->getResponse();
+        }
+
+		// If sessions exist we assume that the responses should not be cached by CDNs / proxies as we are
```

Comment:
Typical from copy/pasting 3.x code... it seems to have copied the indentation too sorry.